tower/src/util/boxed/layer.rs

```diff
@@ -0,0 +1,86 @@
+use crate::util::BoxService;
+use std::{fmt, sync::Arc};
+use tower_layer::{layer_fn, Layer};
+use tower_service::Service;
+
+/// A boxed `Layer` trait object.
+///
+/// [`BoxLayer`] turns a layer into a trait object, allowing the output service to be dynamic.
+///
+/// # Example
+///
+/// `BoxLayer` can for example be useful to create layers dynamically that otherwise wouldn't have
+/// the same types.
+///
+/// ```
+/// use std::time::Duration;
+/// use tower::{Service, ServiceBuilder, BoxError, util::BoxLayer};
+///
+/// fn common_layer<S, T>() -> BoxLayer<S, T, S::Response, BoxError>
+/// where
+///     S: Service<T> + Send + 'static,
+///     S::Future: Send + 'static,
+///     S::Error: Into<BoxError> + 'static,
+/// {
+///     let builder = ServiceBuilder::new()
+///         .concurrency_limit(100);
+///
+///     if std::env::var("SET_TIMEOUT").is_ok() {
+///         let layer = builder
+///             .timeout(Duration::from_secs(30))
+///             .into_inner();
+///
+///         BoxLayer::new(layer)
+///     } else {
+///         let layer = builder
+///             .map_err(Into::into)
+///             .into_inner();
+///
+///         BoxLayer::new(layer)
+///     }
+/// }
+/// ```
+pub struct BoxLayer<In, T, U, E> {
+    boxed: Arc<dyn Layer<In, Service = BoxService<T, U, E>> + Send + Sync + 'static>,
+}
+
+impl<In, T, U, E> BoxLayer<In, T, U, E> {
+    /// Create a new [`BoxLayer`].
+    pub fn new<L>(inner_layer: L) -> Self
+    where
+        L: Layer<In> + Send + Sync + 'static,
+        L::Service: Service<T, Response = U, Error = E> + Send + 'static,
+        <L::Service as Service<T>>::Future: Send + 'static,
+    {
+        let layer = layer_fn(move |inner: In| {
+            let out = inner_layer.layer(inner);
+            BoxService::new(out)
+        });
+
+        Self {
+            boxed: Arc::new(layer),
+        }
+    }
+}
+
+impl<In, T, U, E> Layer<In> for BoxLayer<In, T, U, E> {
+    type Service = BoxService<T, U, E>;
+
+    fn layer(&self, inner: In) -> Self::Service {
+        self.boxed.layer(inner)
+    }
+}
+
+impl<In, T, U, E> Clone for BoxLayer<In, T, U, E> {
+    fn clone(&self) -> Self {
+        Self {
+            boxed: Arc::clone(&self.boxed),
+        }
+    }
+}
+
+impl<In, T, U, E> fmt::Debug for BoxLayer<In, T, U, E> {
+    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
+        fmt.debug_struct("BoxLayer").finish()
```

Comment:
here's a thought; not a blocker for this PR: something we _could_ do, and that we might want to think about doing for `BoxService` as well, is "remembering" the type name of the erased layer, like this:
```rust
pub struct BoxLayer<In, T, U, E> {
    boxed: Arc<dyn Layer<In, Service = BoxService<T, U, E>> + Send + Sync + 'static>,
    name: &'static str,
}

impl<In, T, U, E> BoxLayer<In, T, U, E> {
    /// Create a new [`BoxLayer`].
    pub fn new<L>(inner_layer: L) -> Self
    where
        L: Layer<In> + Send + Sync + 'static,
        L::Service: Service<T, Response = U, Error = E> + Send + 'static,
        <L::Service as Service<T>>::Future: Send + 'static,
    {
        let layer = /* ... */

        Self {
            boxed: Arc::new(layer),
            name: std::any::type_name::<L>(),
        }
    }
}

// ...

impl<In, T, U, E> fmt::Debug for BoxLayer<In, T, U, E> {
    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
        fmt.debug_struct("BoxLayer")
            .field("boxed", &format_args!("Box<{}>", self.name))
            .finish()
     }
}
```

This would let us see what the original, un-erased type was in the `Debug` impl, which could be useful for debugging in some cases.

The main downside to this is that it makes the one-word `BoxLayer` struct suddenly three words long (a `&'static str` is a ptr+len, so two words). This might be meaningful if you're cloning a really large number of `BoxLayer`s. 

Alternatively, we could stick the type name inside the `Arc`, so it's not part of each clone of the struct...but then we'd need some
```rust
struct Inner<In, T, U, E> {
    boxed: Box<dyn Layer<In, Service = BoxService<T, U, E>> + Send + Sync + 'static>,
    name: &'static str,
}
```
which is _also_ not great — now we're dereferencing _two_ heap pointers to call `self.inner.layer`, rather than one.

A third idea is to do the first proposal, but stick `#[cfg(debug_assertions)]` on the `name` field and the code that uses it. That way, type names are included in debug mode, but we remove the extra two words from `BoxLayer` in release mode for performance. IMO, that might be the best choice if we decide to do something like this.

Anyway, just an idea...probably out of scope for this PR and best done in a follow up, if we decide this is worth it.